server/schedule/checker/priority_checker.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2021 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"time"
+
+	"github.com/tikv/pd/pkg/cache"
+	"github.com/tikv/pd/server/config"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule/opt"
+)
+
+// the default value of priority queue size
+const defaultPriorityQueueSize = 1280
+
+// PriorityChecker ensures high priority region should run first
+type PriorityChecker struct {
+	cluster opt.Cluster
+	opts    *config.PersistOptions
+	queue   *cache.PriorityQueue
+}
+
+// NewPriorityChecker creates a priority checker.
+func NewPriorityChecker(cluster opt.Cluster) *PriorityChecker {
+	opts := cluster.GetOpts()
+	return &PriorityChecker{
+		cluster: cluster,
+		opts:    opts,
+		queue:   cache.NewPriorityQueue(defaultPriorityQueueSize),
+	}
+}
+
+// GetType returns PriorityChecker's type
+func (p *PriorityChecker) GetType() string {
+	return "priority-checker"
+}
+
+// RegionPriorityEntry records region priority info
+type RegionPriorityEntry struct {
+	Retry    int
+	Last     time.Time
+	regionID uint64
+}
+
+// ID implement PriorityQueueItem interface
+func (r RegionPriorityEntry) ID() uint64 {
+	return r.regionID
+}
+
+// NewRegionEntry construct of region priority entry
+func NewRegionEntry(regionID uint64) *RegionPriorityEntry {
+	return &RegionPriorityEntry{regionID: regionID, Last: time.Now(), Retry: 1}
+}
+
+// Check check region's replicas, it will put into priority queue if the region lack of replicas.
+func (p *PriorityChecker) Check(region *core.RegionInfo) {
+	makeupCount := 0
+	if p.opts.IsPlacementRulesEnabled() {
+		makeupCount = p.checkRegionInPlacementRule(region)
+	} else {
+		makeupCount = p.checkRegionInReplica(region)
+	}
+	priority := 0 - makeupCount
+	p.addPriorityQueue(priority, region.GetID())
+}
+
+// checkRegionInPlacementRule check region in placement rule mode
+func (p *PriorityChecker) checkRegionInPlacementRule(region *core.RegionInfo) (makeupCount int) {
+	fit := p.cluster.FitRegion(region)
+	if len(fit.RuleFits) == 0 {
+		return
+	}
+	for _, rf := range fit.RuleFits {
+		makeupCount = makeupCount + rf.Rule.Count - len(rf.Peers)
```

Comment:
Is it possible that there will be multiple duplicated rules describing the same replica count here?

Reply:
well, it will also put into priority queue  if there are two rule relate with the same region and one rule not satisfy. then it will make up the deficiency replicas